### PR TITLE
Allow efficient can_match phases on frozen indices

### PR DESCRIPTION
**Pull request by s1monw**
This change adds a special caching reader that caches all relevant
values for a range query to rewrite correctly in a can_match phase
without actually opening the underlying directory reader. This
allows frozen indices to be filtered with can_match and in-turn
searched with wildcards in a efficient way since it allows us to
exclude shards that won't match based on their date-ranges without
opening their directory readers.

Relates to #34352
Depends on #34357

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        }
+
+        @Override
+        public NumericDocValues getNormValues(String field) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public FieldInfos getFieldInfos() {
+            return new FieldInfos(new FieldInfo[0]);
```
Comment (jimczi) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java]: I think you need to handle the point values field here to make `FieldSubsetReader` happy ?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                            }
+
+                            @Override
+                            public long size() {
+                                return size;
+                            }
+
+                            @Override
+                            public int getDocCount() {
+                                throw new UnsupportedOperationException();
```
Comment (jimczi) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java]: Let's add this too ?

**Review by jimczi — COMMENTED** [review]
It looks great @s1monw , I left one comment regarding field security.

**Comment by s1monw** [comment]
@jimczi I pushed changes

**Review by jimczi — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    @Override
+    protected void doClose() {
+
+    }
+
+    @Override
+    public CacheHelper getReaderCacheHelper() {
+        return null;
```
Comment (jpountz) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java]: I believe all above methods should never be called, should we throw an UOE instead for safety?

**Review by jpountz — APPROVED** [review]
Nice way of making can_match work on frozen indices. I left a minor question.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java (hunk 3 above)
Comment (s1monw) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/index/engine/RewriteCachingDirectoryReader.java]: ++

**Pull request merged**